src/webui/api/torrentscontroller.cpp

```diff
@@ -1525,3 +1682,232 @@ void TorrentsController::setSSLParametersAction()
 
     torrent->setSSLParameters(sslParams);
 }
+
+void TorrentsController::metadataAction()
+{
+    const QString sourceParam = params()[u"source"_s].trimmed();
+    // must provide some value to parse
+    if (sourceParam.isEmpty())
+        throw APIError(APIErrorType::BadParams, tr("Must specify URI or hash"));
+
+    QList<QNetworkCookie> cookies;
+    const QString cookie = params()[u"cookie"_s];
+    if (!cookie.isEmpty())
+    {
+        for (QString cookieStr : cookie.split(u"; "_s))
+        {
+            cookieStr = cookieStr.trimmed();
+            int index = cookieStr.indexOf(u'=');
+            if (index > 1)
+            {
+                QByteArray name = cookieStr.left(index).toLatin1();
+                QByteArray value = cookieStr.right(cookieStr.length() - index - 1).toLatin1();
+                cookies += QNetworkCookie(name, value);
+            }
+        }
+    }
+
+    const QUrl sourceUrl = QUrl::fromPercentEncoding(sourceParam.toUtf8());
+    const QString source = sourceUrl.toString();
+
+    const auto iter = m_torrentSource.find(source);
+    // metadata has already been requested
+    if (iter != m_torrentSource.end())
+    {
+        const BitTorrent::InfoHash &infoHash = iter.value();
+        // metadata has already been downloaded
+        if (isMetadataDownloaded(infoHash))
+            setResult(serializeTorrentInfo(m_torrentMetadata.value(infoHash)));
+        else
+        {
+            setStatus(APIStatus::Async);
+            setResult(serializePartialTorrentInfo(infoHash));
+        }
+
+        return;
+    }
+
+    // http(s) url
+    if (Net::DownloadManager::hasSupportedScheme(source))
+    {
+        qDebug("Fetching torrent %s", qUtf8Printable(source));
+        const auto *pref = Preferences::instance();
+
+        if (cookies.size() > 0)
+            Net::DownloadManager::instance()->setCookiesFromUrl(cookies, sourceUrl);
+
+        Net::DownloadManager::instance()->download(Net::DownloadRequest(source).limit(pref->getTorrentFileSizeLimit())
+                , pref->useProxyForGeneralPurposes(), this, &TorrentsController::onDownloadFinished);
+
+        m_torrentSource.insert(source, {});
+
+        setStatus(APIStatus::Async);
+        return;
+    }
+
+    // magnet link or info hash
+    if (const auto parseResult = BitTorrent::TorrentDescriptor::parse(source))
+    {
+        const BitTorrent::TorrentDescriptor &torrentDescr = parseResult.value();
+        const BitTorrent::InfoHash &infoHash = torrentDescr.infoHash();
+
+        auto *session = BitTorrent::Session::instance();
+        const BitTorrent::Torrent *const torrent = session->findTorrent(infoHash);
+        // torrent already exists in transfer list
+        if (torrent)
+        {
+            // torrent's metadata may not have downloaded yet
+            if (torrent->info().isValid())
+                setResult(serializeTorrentInfo(*torrent));
```

Comment:
```suggestion
            if (torrent->info().isValid())
            {
                setResult(serializeTorrentInfo(*torrent));
            }
```

Reply:
I thought we omitted brackets for single line conditionals?
https://github.com/qbittorrent/qBittorrent/blob/master/CODING_GUIDELINES.md#e-single-statement-if-blocks

Reply:
> I thought we omitted brackets for single line conditionals? https://github.com/qbittorrent/qBittorrent/blob/master/CODING_GUIDELINES.md#e-single-statement-if-blocks

You were sloppy when reading it.
>Notice that for a series of `if - else` branches, if one branch needs braces then all branches must add braces.

And code example:
```c++
if (a < b)
{
    do(a);
}
else if (a > b)
{
    // curly braces required here, then all branches should also add them
    do(b);
    do(d);
}
else
{
    do(c);
}
```